src/lib/pipeline.ml

```diff
@@ -261,64 +258,147 @@ module Full (Bfx: Extended_edsl.Semantics) = struct
   type rna_results =
     { rna_bam: [ `Bam ] Bfx.repr;
       stringtie: [ `Gtf ] Bfx.repr;
-      seq2hla: [ `Seq2hla_result ] Bfx.repr option;
-      optitype_rna: [ `Optitype_result ] Bfx.repr option;
       rna_bam_flagstat: [ `Flagstat ] Bfx.repr }
+
   let rna_pipeline
-      ~parameters ~reference_build ~with_seq2hla ~with_optitype_rna samples =
-    let bam = rna_bam ~parameters ~reference_build samples in
-    let fqs = concat_samples samples in
+      ~parameters ~reference_build samples =
+    let bam = to_bam_rna ~parameters ~reference_build samples in
     (* Seq2HLA does not work on mice: *)
-    let seq2hla, optitype_rna =
-      (match reference_build, with_seq2hla, with_optitype_rna with
-     | "mm10", _, _ -> (None, None)
-     | _, false, false -> (None, None)
-     | _, false, true -> (None, Some (optitype_hla fqs `RNA "RNA"))
-     | _, true, false -> (Some (seq2hla_hla fqs), None)
-     | _, true, true -> (Some (seq2hla_hla fqs), Some (optitype_hla fqs `RNA "RNA"))) in
     { rna_bam = bam |> Bfx.save "rna-bam";
       stringtie = bam |> Bfx.stringtie |> Bfx.save "stringtie";
-      rna_bam_flagstat = bam |> Bfx.flagstat |> Bfx.save "rna-bam-flagstat";
-      seq2hla; optitype_rna; }
+      rna_bam_flagstat = bam |> Bfx.flagstat |> Bfx.save "rna-bam-flagstat";}
+
+  type fastqc_results = {
+    normal_fastqcs: [ `Fastqc ] Bfx.repr list;
+    tumor_fastqcs: [ `Fastqc ] Bfx.repr list;
+    rna_fastqcs: [ `Fastqc ] Bfx.repr list option; }
+  let fastqc_pipeline ~normal_samples ~tumor_samples ?rna_samples () =
+    let normal_fastqcs =
+      List.mapi
+        ~f:(fun i fq -> qc fq |> Bfx.save (sprintf "QC:normal-%d" i))
+        normal_samples
+    in
+    let tumor_fastqcs =
+      List.mapi
+        ~f:(fun i fq -> qc fq |> Bfx.save (sprintf "QC:tumor-%d" i))
+        tumor_samples
+    in
+    let rna_fastqcs =
+      match rna_samples with
+      | None -> None
+      | Some rna_samples ->
+        Some
+          (List.mapi
+             ~f:(fun i fq -> qc fq |> Bfx.save (sprintf "QC:rna-%d" i))
+             rna_samples)
+    in
+    { normal_fastqcs; tumor_fastqcs; rna_fastqcs }
+
+
+  let email_pipeline
+      ?rna_results ~parameters ~normal_bam_flagstat ~tumor_bam_flagstat ~fastqcs
+    =
+    let rna_bam_flagstat =
+      rna_results >>= fun {rna_bam_flagstat; _} -> return rna_bam_flagstat in
```

Comment:
The `open Option` is too far up,` `Option` is not meant to be opened globally (Esp. since this case is actually an `Option.map` rather than `bind`).